### ISSUE TITLE
Determining effective training

### ISSUE THREAD

**Issue by dboshardy** [yuval-alaluf/restyle-encoder#64]
I'm trying to train a model based off a stylegan i trained similar to toonify.  

I'm attempting to train the restyle-pSp model on this dataset and ffhq for the translation task.  However, after quite a long time (can't set batch size at 8 due to GPU memory running out despite running with a 16GB of VRAM), the loss isn't really changing (see below):

I'm training using the following options:
```yaml
batch_size: 6
board_interval: 50
checkpoint_path: null
dataset_type: "my_ffhq_dataset"
encoder_type: "BackboneEncoder"
exp_dir: "experiment/restyle_psp"
id_lambda: 1
image_interval: 100
input_nc: 6
l2_lambda: 1
learning_rate: 0.0001
lpips_lambda: 0.8
max_steps: 500000
moco_lambda: 0
n_iters_per_batch: 5
optim_name: "ranger"
output_size: 1024
save_interval: 2000
start_from_latent_avg: true
stylegan_weights: "../models/model.pt"
test_batch_size: 2
test_workers: 2
train_decoder: false
val_interval: 1000
w_norm_lambda: 0.25
workers: 4
```

The below metrics are what they looked like after only 1k iterations. And looking at the model output in the experiment dir, the output faces aren't changing in the slightest.  The source and target are appropriately configured as they are the normal and the toon version of the normal face.  It's a dataset of 1024x1024 in 3500 pairs for training and 1500 for testing.
```shell
Metrics for train, step 20900
        loss_id =  0.44862064719200134
        id_improve =  0.47951146960258484
        loss_l2 =  0.09637744724750519
        loss_lpips =  0.4569443166255951
        loss_w_norm =  0.0013350744266062975
        loss =  0.9108873605728149
Metrics for train, step 20950
        loss_id =  0.5060080885887146
        id_improve =  0.322285289876163
        loss_l2 =  0.09358177334070206
        loss_lpips =  0.4740564227104187
        loss_w_norm =  0.0016432201955467463
        loss =  0.9792457818984985
Metrics for train, step 21000
        loss_id =  0.4523411691188812
        id_improve =  0.38620127489169437
        loss_l2 =  0.12004581093788147
        loss_lpips =  0.44153523445129395
        loss_w_norm =  0.0012316436041146517
        loss =  0.9259231090545654
```

**Comment by yuval-alaluf** [comment]
Any chance you can add an output image that you got during training? 
In addition, I always recommend trying to overfit on a very small set of images (e.g., 10) to make sure that everything is configured properly.

**Comment by dboshardy** [comment]
@yuval-alaluf unfortunately I'm not able to upload the images.  But I can say that the output is perceptually nearly identical to the starting latent average.  And the outputs' Target Sim (what is this metric, by the way?) never strays from 0.47 to 0.63 or so.  I'll try the overfitting on a smaller dataset.  Perhaps the IRSE50 model is unfit for analyzing face-adjacent images if they stray too far from human faces?

I tried looking in `train_options.py`, but is there any explanation of what the various options are doing for training? Some are more intuitive than others.

I also wanted to ask, you mention in the Encoder Bootstrapping section that the toonify version is trained on real images and not using paired images.   I'm not sure I understand how that would work?  Does that mean that the toonify generator was trained using real images that had been projected instead of random latents and then those real/toon pairs were used to train restyle?  Or was it something else?

**Comment by yuval-alaluf** [comment]
> Perhaps the IRSE50 model is unfit for analyzing face-adjacent images if they stray too far from human faces?

If I understood correctly, you're translating from real faces to toons, so it's definitely possible that the ArcFace loss isn't very useful here, but it's still quite surprising that no change whatsoever is made. I don't think that the IDLoss is the problem here. I think the overfitting experiment will help clear up if there are other issues here so I'm looking forward to an update on that experiment. 

Regarding the training options, if there are specific things that are unclear please let me know and I'll clarify. 

Regarding your final question: when we say that the toonify version was trained on real images, it means that the encoder was trained only on real images. That is, the generator is trained to output toons and we train the encoder using only real images (i.e., no paired data). The toon generator itself was trained using Justin Pinkney's layer swapping technique (you can find the paper [here](https://arxiv.org/abs/2010.05334).

**Comment by dboshardy** [comment]
Finally had a chance to come back to this.  It trained super well on the overfit data (100/50 imgs train/test).  Training on 5k/1k split now and it's kind of oscillating in the loss metric.  It's certainly making progress. Some images in the train and test dirs in the experiment dir are very close to the target but some have only just started to diverge from the latent average. Is this expected?  I'm tempted to keep letting it run for another day (had to lower the batch size to 4 as I only have the lower memory V100). I imagine this has to do with how far various target images are from the latent average? Does one step include every image in the training set, or only a subset?
